File: tradebot/strategy.py
"""Rule-based RSI baseline.

Used as a sanity check and as the default strategy when the user runs
``backtest`` or ``paper`` without passing ``--agent``.
"""
from __future__ import annotations

from dataclasses import dataclass

from .env import BUY_CALL, HOLD, SELL, OptionsEnv, State


@dataclass
class RSIStrategy:
    rsi_buy_below: float = 35.0
    rsi_sell_above: float = 65.0
    take_profit_pct: float = 0.25
    stop_loss_pct: float = 0.20

    def act(self, state: State, env: OptionsEnv) -> int:
        # Our discrete RSI bins are [low, mid_low, mid_high, high] per config.
        # We also look at the raw values on the env so the thresholds apply.
        i = min(env.t, len(env._close) - 1)
        rsi_val = float(env._rsi[i])
        has_pos = state[2] == 1

        if has_pos:
            # Check take-profit / stop-loss on first open position.
            pos = env.broker.positions[0] if env.broker and env.broker.positions else None
            if pos is not None:
                spot = float(env._close[i])
                mark = env._mark(pos, env._dates[i].date(), spot)
                pnl_pct = (mark - pos.entry_price) / max(pos.entry_price, 1e-6)
                if pnl_pct >= self.take_profit_pct or pnl_pct <= -self.stop_loss_pct:
                    return SELL
            if rsi_val >= self.rsi_sell_above:
                return SELL
            return HOLD

        if rsi_val <= self.rsi_buy_below:
            return BUY_CALL
        return HOLD

File: lib/strategy.ts
// Rule-based RSI baseline (TS port of tradebot/strategy.py).

import { BUY_CALL, HOLD, OptionsEnv, SELL, type State } from "./env";

export interface RsiStrategyOptions {
  rsiBuyBelow?: number;
  rsiSellAbove?: number;
  takeProfitPct?: number;
  stopLossPct?: number;
}

export class RsiStrategy {
  rsiBuyBelow: number;
  rsiSellAbove: number;
  takeProfitPct: number;
  stopLossPct: number;

  constructor(opts: RsiStrategyOptions = {}) {
    this.rsiBuyBelow = opts.rsiBuyBelow ?? 35;
    this.rsiSellAbove = opts.rsiSellAbove ?? 65;
    this.takeProfitPct = opts.takeProfitPct ?? 0.25;
    this.stopLossPct = opts.stopLossPct ?? 0.2;
  }

  act(_state: State, env: OptionsEnv): number {
    const i = Math.min(env.t, env.series.bars.length - 1);
    const rsi = env.indicators.rsi[i];
    const hasPos = env.broker.hasPosition();

    if (hasPos) {
      const pos = env.broker.positions[0];
      const spot = env.series.bars[i].close;
      const mark = env.mark(pos, env.series.bars[i].t, spot);
      const pnlPct = (mark - pos.entryPrice) / Math.max(pos.entryPrice, 1e-6);
      if (pnlPct >= this.takeProfitPct || pnlPct <= -this.stopLossPct) {
        return SELL;
      }
      if (rsi >= this.rsiSellAbove) return SELL;
      return HOLD;
    }

    if (rsi <= this.rsiBuyBelow) return BUY_CALL;
    return HOLD;
  }
}
